ouroboros-consensus-cardano/src/shelley/Ouroboros/Consensus/Shelley/Ledger/Ledger.hs

```diff
@@ -338,49 +360,56 @@ instance ShelleyCompatible proto era
   --    - 'updateChainDepState': executes the @PRTCL@ transition
   -- + 'applyBlockLedgerResult': executes the @BBODY@ transition
   --
-  applyBlockLedgerResult =
-      applyHelper (swizzle ..: appBlk)
+  applyBlockLedgerResultWithOpts opts cfg = withEvents (events opts) $ \ep ->
+      applyHelper (swizzle ep ..: appBlk ep) cfg
     where
-      swizzle m =
+      swizzle ::
+           STS.SingEP ep
+        -> Except
+             (SL.BlockTransitionError era)
+             (STS.EventReturnType ep (Core.EraRule "BBODY" era) (SL.NewEpochState era))
+        -> Except
+             (ShelleyLedgerError era)
+             (LedgerResult (LedgerState (ShelleyBlock proto era)) (SL.NewEpochState era))
+      swizzle STS.EPDiscard m =
+        withExcept BBodyError m <&> \l ->
+          LedgerResult {
+              lrEvents = []
+            , lrResult = l
+            }
+      swizzle STS.EPReturn m =
         withExcept BBodyError m <&> \(l, events) ->
           LedgerResult {
               lrEvents = map ShelleyLedgerEventBBODY events
             , lrResult = l
             }
 
+
       -- Apply the BBODY transition using the ticked state
-      appBlk =
+      appBlk ::
+           STS.EventReturnTypeRep ep
+        => STS.SingEP ep
+        -> SL.Globals
+        -> SL.NewEpochState era
+        -> SL.Block (SL.BHeaderView (ProtoCrypto proto)) era
+        -> Except
+             (SL.BlockTransitionError era)
+             (STS.EventReturnType ep (Core.EraRule "BBODY" era) (SL.NewEpochState era))
+      appBlk ep' =
         SL.applyBlockOpts
           STS.ApplySTSOpts {
               asoAssertions = STS.globalAssertionPolicy
-            , asoValidation = STS.ValidateAll
-            , asoEvents     = STS.EPReturn
+            , asoValidation = validationMode (validation opts)
```

Comment:
Same question: we had some hard-coded options, why are we changing them?

Reply:
This line is exactly the key to this PR. Now you can choose which options are passed to `applyBlockLedgerResult` and it is at this point where they are set.

Reply:
Right, makes sense, thanks!